### PR TITLE
Added AreaRouting and Sharedconfig for sending emails for contact adv…

### PR DESCRIPTION
**Pull request by kaderhashmi**
…isor

### CODE REVIEW

File: DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            {
+                var template = emailTemplateRepository.GetByTemplateName(sendEmailRequest.TemplateName);
+
+                try
+                {
+                    var url = $"{ConfigurationManager.AppSettings[Constants.AreaRoutingApiServiceUrl]}?location={sendEmailRequest.Postcode}";
+                    var accessKey = ConfigurationManager.AppSettings[Constants.AreaRoutingApiSubscriptionKey];
+
+                    httpClientService.AddHeader(Constants.OcpApimSubscriptionKey, accessKey);
+                    var response = await this.httpClientService.GetWhereAsync(url, (httpResponseMessage) => !httpResponseMessage.IsSuccessStatusCode && httpResponseMessage.StatusCode != HttpStatusCode.NoContent);
```
Comment (Muthuramana) [inline DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs]: Can this not be an polymorphic method of GetAsync

File: DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                    var areaRoutingApiResponse = await response.Content.ReadAsAsync<AreaRoutingApiResponse>();
+
+                    template.To = areaRoutingApiResponse.EmailAddress;
+                }
+                catch (BrokenCircuitException)
```
Comment (Muthuramana) [inline DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs]: Check this exception was logged before getting here if not, then use IApplicationLogger and use ErrorJustLogIt

File: DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                {
+                    var fallbackResponse = await sharedConfigurationService.GetConfigAsync<AreaRoutingApiResponse>(SharedConfigServiceName, SharedConfigKeyName);
+                    template.To = fallbackResponse.EmailAddress;
+                }
+
+                return await SendEmail(sendEmailRequest, template).ConfigureAwait(false);
```
Comment (Muthuramana) [inline DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs]: Can we use SendEmailAsync with optional parameter for email template?

File: DFC.Digital/DFC.Digital.Services.SendGrid/SendGridEmailService.cs [hunk 4]
```diff
@@ -34,31 +34,36 @@ public class SendGridEmailService : INoncitizenEmailService<ContactUsRequest>
...
         }
 
-        public async Task<bool> SendEmailAsync(ContactUsRequest sendEmailRequest)
+        public virtual async Task<bool> SendEmailAsync(ContactUsRequest sendEmailRequest)
+        {
+            var template = emailTemplateRepository.GetByTemplateName(sendEmailRequest.TemplateName);
+            return await SendEmail(sendEmailRequest, template).ConfigureAwait(false);
+        }
+
+        protected async Task<bool> SendEmail(ContactUsRequest sendEmailRequest, EmailTemplate template)
```
Comment (Muthuramana) [inline DFC.Digital/DFC.Digital.Services.SendGrid/SendGridEmailService.cs]: Can we change this so SendEmailAsync can accept email template as optional param

```
 public virtual async Task<bool> SendEmailAsync(ContactUsRequest sendEmailRequest, EmailTemplate template = null)
        {
            if (simulateEmailResponsesService.IsThisSimulationRequest(sendEmailRequest.Email))
            {
                return simulateEmailResponsesService.SimulateEmailResponse(sendEmailRequest.Email);
            }

            if (template == null)
            {
                template = emailTemplateRepository.GetByTemplateName(sendEmailRequest.TemplateName);
            }

           
```

File: DFC.Digital/DFC.Digital.Web.Sitefinity/web.config.template [hunk 5]
```diff
@@ -87,6 +87,16 @@
...
 
+    <!--Area Routing API settings-->
+    <add key="DFC.Digital.AreaRoutingApi.ServiceUrl" value="__AreaRoutingAPIAreasEndpoint__" />
+    <add key="DFC.Digital.AreaRoutingApi.SubscriptionKey" value="__AreaRoutingApiKey__" />
+
+    <!--SharedConfig Settings-->
+    <add key="DFC.SharedConfig.ConfigurationStorageConnectionString" value="__SharedConfigStorageConnectionString__" />
+    <add key="DFC.SharedConfig.CloudStorageTableName" value="Configuration" />
+    <add key="DFC.SharedConfig.InMemoryCacheTimeToLiveTimespan" value="00:01:00" />
+    <add key="DFC.SharedConfig.EnvironmentName" value="Production" />
```
Comment (Muthuramana) [inline DFC.Digital/DFC.Digital.Web.Sitefinity/web.config.template]: Remove these as defaults are sufficient

File: DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            ISharedConfigurationService sharedConfigurationService) : base(emailTemplateRepository, mergeEmailContentService, auditRepository, simulateEmailResponsesService, sendGridClient, mapper)
+        {
+            this.emailTemplateRepository = emailTemplateRepository;
+            this.httpClientService = httpClientService;
+            this.sharedConfigurationService = sharedConfigurationService;
+        }
+
+        public override async Task<bool> SendEmailAsync(ContactUsRequest sendEmailRequest)
+        {
+            if (sendEmailRequest.ContactOption == "ContactAdviser")
```
[the remark this answers is not in the document]
Comment (kaderhashmi) [reply DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs]: Updated as suggested - required moving the ContactOption enum to data project

File: DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            {
+                var template = emailTemplateRepository.GetByTemplateName(sendEmailRequest.TemplateName);
+
+                try
+                {
+                    var url = $"{ConfigurationManager.AppSettings[Constants.AreaRoutingApiServiceUrl]}?location={sendEmailRequest.Postcode}";
```
[the remark this answers is not in the document]
Comment (kaderhashmi) [reply DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs]: Updated as suggested

File: DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs (hunk 1 above)
Comment (kaderhashmi) [reply DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs]: Updated as suggested

File: DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs (hunk 3 above)
Comment (kaderhashmi) [reply DFC.Digital/DFC.Digital.Services.SendGrid/FAMSendGridEmailService.cs]: Updated as suggested

File: DFC.Digital/DFC.Digital.Services.SendGrid/SendGridEmailService.cs (hunk 4 above)
Comment (kaderhashmi) [reply DFC.Digital/DFC.Digital.Services.SendGrid/SendGridEmailService.cs]: Updated as suggested

File: DFC.Digital/DFC.Digital.Web.Sitefinity/web.config.template (hunk 5 above)
Comment (kaderhashmi) [reply DFC.Digital/DFC.Digital.Web.Sitefinity/web.config.template]: Removed from template

**Pull request merged**